File: p72.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int n = word1.size() + 1;
        int m = word2.size() + 1;
        int result[n][m];
        memset(result, 0, sizeof(result));
        for(int i = 0; i < m; ++i) {
            result[0][i] = i;
        }
        for(int j = 0; j < n; ++j) {
            result[j][0] = j;
        }
        for(int k = 1; k < n; ++k) {
            for(int l = 1; l < m; ++l) {
                result[k][l] = min(result[k-1][l], result[k][l-1]) + 1;
                if(word1[k-1] == word2[l-1]) {
                    result[k][l] = min(result[k][l], result[k-1][l-1]);
                } else {
                    result[k][l] = min(result[k][l], result[k-1][l-1] + 1);
                }
            }
        }
        return result[n-1][m-1];
    }
};

File: p72.py
from functools import lru_cache

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if len(word1) == 0 or len(word2) == 0:
            return max(len(word1), len(word2))

        e1, e2 = len(word1), len(word2)

        @lru_cache(None)
        def dfs(s1, s2):
            if e1-s1 == 0 or e2-s2 == 0:
                return max(e1-s1, e2-s2)

            if word1[s1] == word2[s2]:
                return dfs(s1+1, s2+1)
            return 1 + min(dfs(s1+1, s2), dfs(s1, s2+1), dfs(s1+1, s2+1))

        return dfs(0, 0)
